File: app/app/src/main/java/dev/steamvault/app/ui/library/logic/CoverArt.kt
package dev.steamvault.app.ui.library.logic

/**
 * Cover-art URL + deterministic fallback tile (WP 4b.4) — Kotlin port of
 * `web/js/lib/cover-art.js` (docs/design/vault-app-mockup-NOTES.md,
 * "Covers"): 2:3 portrait artwork, and when it can't be shown (offline, CDN
 * miss for an unlisted appid), a styled tile rather than a blank rectangle.
 * `fallbackHues`/`fallbackPattern` derive their two ingredients
 * deterministically FROM the appid, same as the web port — same game
 * always gets the same fallback look, no extra request, no randomness to
 * make the JVM tests (`CoverArtTest`) flaky. Values are pinned to match the
 * web port's FNV-1a hash byte-for-byte (same inputs, same outputs on both
 * frontends) even though nothing currently cross-checks the two beyond
 * this shared design intent.
 *
 * `library_600x900.jpg` is Valve's own portrait-capsule asset path -- same
 * 2:3 ratio the mockup's fake covers already used. Loading is done with
 * Coil (`ui/library/CoverArtImage.kt`) rather than a hand-rolled bitmap
 * fetch -- see that file's kdoc for the "Coil pinned vs hand-rolled"
 * justification recorded for the reviewer.
 */
const val STEAM_CDN_HOST: String = "cdn.akamai.steamstatic.com"

fun coverArtUrl(appid: Int): String = "https://$STEAM_CDN_HOST/steam/apps/$appid/library_600x900.jpg"

/** Tiny deterministic string hash (FNV-1a), good enough for "pick a stable
 * decorative value from an integer id" -- not a security primitive. Mirrors
 * `web/js/lib/cover-art.js`'s `fnv1a` (32-bit unsigned arithmetic, same as
 * JS's `>>> 0`). */
private fun fnv1a(n: Long): Long {
    var h = 0x811c9dc5L
    for (ch in n.toString()) {
        h = h xor ch.code.toLong()
        h = (h * 0x01000193L) and 0xFFFFFFFFL
    }
    return h and 0xFFFFFFFFL
}

data class FallbackHues(val h1: Int, val h2: Int)

/** Two hues (0-359) for the fallback tile's gradient. Always the same pair
 * for the same appid. */
fun fallbackHues(appid: Int): FallbackHues {
    val h = fnv1a(appid.toLong())
    val h1 = (h % 360).toInt()
    val h2 = ((h ushr 9) % 360).toInt()
    return FallbackHues(h1, h2)
}

/** Which of the mockup's six decorative overlay patterns (0..5) this appid's
 * fallback tile uses. */
fun fallbackPattern(appid: Int): Int {
    // JS: fnv1a(appid * 2654435761) % 6 -- the multiplication is float-lossy
    // in JS for large appids but harmless (it's a decorative hash input,
    // not a security or correctness value); here it is done in Long with an
    // explicit 32-bit mask so the result stays deterministic and in-range
    // without relying on overflow semantics differing between platforms.
    val product = (appid.toLong() * 2654435761L) and 0xFFFFFFFFL
    return (fnv1a(product) % 6L).toInt()
}

File: web/js/lib/cover-art.js
/**
 * Cover-art URL + deterministic fallback tile (WP 4a.3).
 *
 * Ports the mockup's capsule-art approach (docs/design/
 * vault-app-mockup-NOTES.md, "Covers"): 2:3 portrait artwork, and when it
 * can't be shown, a styled tile rather than a blank rectangle. The mockup
 * used per-title hues from its fake seed data (`g.h1`/`g.h2`); the real app
 * has real appids and no seed hues, so `fallbackHues`/`fallbackPattern`
 * below derive the same two ingredients deterministically FROM the appid —
 * same game always gets the same fallback look, no server round trip, no
 * randomness to make headless-test assertions flaky.
 *
 * **CSP note (deliberate extension — see api/vault_api/webui.py and
 * api/README.md's "Security headers" section for the paired change this WP
 * makes there).** `STEAM_CDN_HOST` is the exact, single host added to
 * `img-src`; this module is the ONLY place that constructs a URL against
 * it, so the CSP's scope and this module's scope are the same one line.
 * `library_600x900.jpg` is Valve's own portrait-capsule asset path — same
 * 2:3 ratio the mockup's fake covers already used (NOTES.md: "costs nothing
 * to swap in real Steam artwork later").
 *
 * Pure — no DOM, no fetch — so the URL shape and the fallback hashing are
 * unit-tested headlessly (web/tests/cover-art.test.js); the actual `<img>`
 * element + its onerror-swap-to-fallback wiring lives in
 * components/game-card.js, which is DOM-only and not unit-tested the same
 * way (mirrors the existing split between status-icon.js's pure glyph
 * tables and its DOM builder).
 *
 * **`headerArtUrl` (WP 4h.3).** Same host, same "this module is the ONLY
 * place that builds a URL against it" discipline, different asset:
 * `header.jpg` is Valve's wide 460x215 landscape capsule, used as hero/
 * header art at the top of the game detail card
 * (components/header-art.js) rather than the portrait grid/mini-cover
 * shape above. No CSP change needed — `STEAM_CDN_HOST` is already the
 * exact, single host `img-src` allows (api/vault_api/webui.py), and this
 * is a different PATH on the same host, not a new one.
 */

export const STEAM_CDN_HOST = "cdn.akamai.steamstatic.com";

/** @param {number} appid */
export function coverArtUrl(appid) {
  return `https://${STEAM_CDN_HOST}/steam/apps/${appid}/library_600x900.jpg`;
}

/** Wide (460x215) header/hero capsule art for the game detail card
 * (WP 4h.3). @param {number} appid */
export function headerArtUrl(appid) {
  return `https://${STEAM_CDN_HOST}/steam/apps/${appid}/header.jpg`;
}

/** Tiny deterministic string hash (FNV-1a), good enough for "pick a stable
 * decorative value from an integer id" — not a security primitive. */
function fnv1a(n) {
  let h = 0x811c9dc5;
  const str = String(n);
  for (let i = 0; i < str.length; i++) {
    h ^= str.charCodeAt(i);
    h = Math.imul(h, 0x01000193);
  }
  return h >>> 0;
}

/** Two hues (0-359) for the fallback tile's gradient, mirroring the
 * mockup's `--h1`/`--h2` custom properties. Always the same pair for the
 * same appid. */
export function fallbackHues(appid) {
  const h = fnv1a(appid);
  const h1 = h % 360;
  const h2 = (h >>> 9) % 360;
  return { h1, h2 };
}

/** Which of the mockup's six decorative overlay patterns (p0..p5,
 * css/app.css) this appid's fallback tile uses. */
export function fallbackPattern(appid) {
  return fnv1a(appid * 2654435761) % 6;
}
